sympy/core/operations.py

```diff
@@ -504,3 +505,117 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(OpDispatcher):
+    """
+    Handler dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    Two informations need to be registered. The first is the handler, which is registered
+    to a type and returns the result of operation. The second is the priority relation,
+    which is registered to two types and returns the one with higher priority. Order of types
+    does not matter here because both order are searched when dispatching.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, MatAdd, Expr, MatrixExpr
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> add = AssocOpDispatcher("add", Add)
+
+    >>> @add.register_handler(MatrixExpr)
+    ... def _(*args, **kwargs):
+    ...     return MatAdd(*args, **kwargs)
```

Comment:
We probably shouldn't do this in the doctests as it can interfere with subsequent doctests.

Reply:
I think this should be ok because `add` is a local dispatcher?

Reply:
Got it. Will the examples in `test_operations` raise problems, too? I can apply `add` in other module (like `matrices`) and add See Also section here to refer that case.

Reply:
Oh yes. If the `add` here is different from the one actually used in the `add.py` module then this should be fine

Reply:
Then I'll mark this as resolved.